### PR TITLE
fix: btc destination input

### PR DESCRIPTION
**Pull request by MaxwellDG**
This was initially done in response to #2298 

There were several other minor UI improvements to be made, and 1 large refactoring done to address a lag issue on this input component. I'll reiterate the changes here:
  - There was a 1px jump from borderWidth on pasting
  - Buttons didn't have opacity indication when being pressed
  - Added a toast so users know what's happening when they press the paste button (I was confused when I was trying to press it and I didn't have anything in my clipboard. Initially thought it was broken)
  - Pressing the paste button twice was messing up the validation process and removing it on 2nd press (it went into manual 'entering' mode)
  - theres a multiple seconds lag on pressing the paste button for the first time which is very uncomfortable. Gonna work on making that more snappy. Likely it was only happening on Android
  
  Before merging, I'd like @krtk6160 to confirm if he sees the main issue already resolved here

### CODE REVIEW

**Comment by nicolasburtey** [comment]
there are some prettier issue with this PR

**Comment by nicolasburtey** [comment]
> Before merging, I'd like @krtk6160 to confirm if he sees the main issue already resolved here

I don't think @krtk6160 have the set up ready to test a mobile PR locally easily. I think best for me would be to review the PR and then have @krtk6160 test in on a nightly build after it has been merged. but let me know if you think this is ready for review.

**Comment by MaxwellDG** [comment]
> > Before merging, I'd like @krtk6160 to confirm if he sees the main issue already resolved here
> 
> I don't think @krtk6160 have the set up ready to test a mobile PR locally easily. I think best for me would be to review the PR and then have @krtk6160 test in on a nightly build after it has been merged. but let me know if you think this is ready for review.

Sure, whatever you think is best.

It's ready for review now, and here's a video I took on an Android G20. I'm still trying to get XCode to get the project running so I can test on iOS, but I'll go work on that now while you're testing the PR.


https://github.com/GaloyMoney/galoy-mobile/assets/51105802/ebb820a8-5759-4014-b17e-cd333b25c1a1

**Comment by nicolasburtey** [comment]
still seeing the same behavior by checking out from this PR

https://github.com/GaloyMoney/galoy-mobile/assets/11156370/ca42f5d2-285b-4aa5-bc4d-87541352f595

**Comment by nicolasburtey** [comment]
this is the invoice I used: lnbc10n1pjh09c7pp5cxgqx8ptspl28s43pxt20aae40pcqa3fx86uyyzng2vhup8j0uesdq5g9kxy7fqd9h8vmmfvdjscqzzsxqyz5vqsp5dj42yuzldlf0r4yee5gn3qlctq8tqkp8jqrt77yc38vdgnmeksks9qyyssqt558gadxwp95y0f75n6hrz5mgtf3wtg93xz04jmf4tuatyh8kpyn66vs20ug64xwcwxqsmxe4hjer2utlkjx0fnhpvndvlqlqemq94gq6ecx7l

**Comment by MaxwellDG** [comment]
> still seeing the same behavior by checking out from this PR
> 
>  Screen.Recording.2023-12-11.at.5.12.49.PM.mov

I think I'm extremely confused on what's intended here.

The user wants double-tap to select the entire string right? In both your video, and @krtk6160 video I see you guys double tapping and then the entire string is getting selected. Could you rephrase a description of the behaviour you think is lacking?

**Comment by MaxwellDG** [comment]
> this is the invoice I used: lnbc10n1pjh09c7pp5cxgqx8ptspl28s43pxt20aae40pcqa3fx86uyyzng2vhup8j0uesdq5g9kxy7fqd9h8vmmfvdjscqzzsxqyz5vqsp5dj42yuzldlf0r4yee5gn3qlctq8tqkp8jqrt77yc38vdgnmeksks9qyyssqt558gadxwp95y0f75n6hrz5mgtf3wtg93xz04jmf4tuatyh8kpyn66vs20ug64xwcwxqsmxe4hjer2utlkjx0fnhpvndvlqlqemq94gq6ecx7l
> 
> after I deleted the first part, remained this part: lnbc10n1pjh09c7pp5cxgqx8ptspl28s43pxt20aae40pcqa3fx86uyyzng2vhup8j0uesdq5g9kxy7fqd9h8vmmfvdjscqzzsxqyz

OHHHH okay so it's not working when the string is super long. I'll look into that

**Comment by MaxwellDG** [comment]
> lnbc10n1pjh09c7pp5cxgqx8ptspl28s43pxt20aae40pcqa3fx86uyyzng2vhup8j0uesdq5g9kxy7fqd9h8vmmfvdjscqzzsxqyz5vqsp5dj42yuzldlf0r4yee5gn3qlctq8tqkp8jqrt77yc38vdgnmeksks9qyyssqt558gadxwp95y0f75n6hrz5mgtf3wtg93xz04jmf4tuatyh8kpyn66vs20ug64xwcwxqsmxe4hjer2utlkjx0fnhpvndvlqlqemq94gq6ecx7l

Found the prop that fixes that. Ready for a real review now.

**Comment by nicolasburtey** [comment]
> theres a multiple seconds lag on pressing the paste button for the first time which is very uncomfortable. Gonna work on making that more snappy. Likely it was only happening on Android

I have not seen this behavior AFAIR, maybe expect with lnurl? for what type of payment method did you see this behavior?

the updates you made seems great given the highlighted issue, but that also make a code a bit more complex to comprehend, so I'm wondering if we don't have an issue otherwise in our code. I don't think it should take long to validate an invoice for instance

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx [hunk 1]
```diff
@@ -102,36 +130,45 @@ const SendBitcoinDestinationScreen: React.FC<Props> = ({ route }) => {
...
+    dispatchDestinationStateAction({
+      type: SendBitcoinActions.SetValidating,
+      payload: {},
+    })
+    return true
+  }, [bitcoinNetwork, wallets, contacts])
+
+  const validateDestination = React.useCallback(
+    async (rawInput: string, obj: NetworkWalletsContacts) => {
+      // extra check to ensure nothing has changed since initiating validation
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:162]: maybe I'm missing something, but in the line below, we are just checking for presence of value, not that the value has changed?

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx [hunk 2]
```diff
@@ -240,14 +271,37 @@ const SendBitcoinDestinationScreen: React.FC<Props> = ({ route }) => {
     }
   }, [destinationState, goToNextScreenWhenValid, navigation, setGoToNextScreenWhenValid])
 
-  const initiateGoToNextScreen = useMemo(() => {
-    return async () => {
-      if (validateDestination) {
-        await validateDestination(destinationState.unparsedDestination)
-        setGoToNextScreenWhenValid(true)
+  // setTimeout here allows for the main JS thread to update the UI before the long validateDestination call
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:274]: how would the UI been updated? Is it the opacity of the button that would react for instance? or is it also that the text should be shown while it's been validated?

**Comment by nicolasburtey** [comment]
it seems there is no automatic validation on paste with a "right click" long button press in the input field. but only on the paste button. 

Uploading Screen Recording 2023-12-11 at 6.05.54 PM.mov…

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx [hunk 3]
```diff
@@ -240,14 +271,37 @@ const SendBitcoinDestinationScreen: React.FC<Props> = ({ route }) => {
...
+        returnValidateDestinationFunc(input, {
+          _bitcoinNetwork: bitcoinNetwork,
+          _wallets: wallets,
+          _contacts: contacts,
+        }),
+        0,
+      )
+    },
+    [bitcoinNetwork, wallets, contacts, validateDestination],
+  )
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:293]:
```suggestion
  const waitAndValidateDestination = React.useCallback(
    (input: string) => {
      setTimeout(
        () =>
          validateDestination(input, {
            _bitcoinNetwork: bitcoinNetwork,
            _wallets: wallets,
            _contacts: contacts,
          }),
        0,
      )
    },
    [bitcoinNetwork, wallets, contacts, validateDestination],
  )
```

I dont think `returnValidateDestinationFunc` is really necessary

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx [hunk 4]
```diff
@@ -65,6 +66,33 @@ gql`
...
+      readonly __typename: "UserContact"
+      readonly id: string
+      readonly username: string
+    }[]
+  | undefined
+
+type NetworkWalletsContacts = {
+  _bitcoinNetwork: Network | undefined
+  _wallets: Wallets
+  _contacts: Contacts
```
Comment (nicolasburtey) [inline app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:93]: network, wallets and contacts are immutable value in the context of this flow. do we need to move those values around? It's not clear to me why we would want to do that or what would be the benefits of it

Comment (nicolasburtey) [reply app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:93]: * immutable as is: they won't change, unless if there were out of sync and will be re-sync with the query to the backend server when entering the screen.

and really contacts it the only one that could be updated following a payment. those other values are value that won't change for a particular user, so they would only change after a log out / login with new user or login to new environment.

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx (hunk 1 above)
Comment (MaxwellDG) [reply app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:162]: You're correct it's just for the presence of value. I pass them as such because of typescript annoyances. In the validateDestination function the values are used a few times and they must be verified to not be undefined (since the type Network that comes from GraphQL for example has the type 'Network | undefined').

**Comment by MaxwellDG** [comment]
> > theres a multiple seconds lag on pressing the paste button for the first time which is very uncomfortable. Gonna work on making that more snappy. Likely it was only happening on Android
> 
> I have not seen this behavior AFAIR, maybe expect with lnurl? for what type of payment method did you see this behavior?
> 
> the updates you made seems great given the highlighted issue, but that also make a code a bit more complex to comprehend, so I'm wondering if we don't have an issue otherwise in our code. I don't think it should take long to validate an invoice for instance



> it seems there is no automatic validation on paste with a "right click"/long button press in the input field. but only on the paste button.
> 
>  Screen.Recording.2023-12-11.at.6.05.54.PM.mov

Yea this is expected behaviour. It was like that before. It's cause it's considering it as text being input from the keyboard. I could change it to automatically run a validation after being pasted if you'd like

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx (hunk 3 above)
Comment (MaxwellDG) [reply app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:293]: ehhhh maybe. I was having difficulty with variable input being stale when sent through. I'll see if I can remove that part, it could've been overkill.

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx (hunk 4 above)
Comment (MaxwellDG) [reply app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:93]: Didn't know they were immutable. Saw them coming in from GraphQL queries, and they weren't really being handled as such in the rest of the code. If they're immutable though you're right I could remove them from the setTimeout -> validateDestination functions and greatly simplify the code.

**Comment by MaxwellDG** [comment]
> > theres a multiple seconds lag on pressing the paste button for the first time which is very uncomfortable. Gonna work on making that more snappy. Likely it was only happening on Android
> 
> I have not seen this behavior AFAIR, maybe expect with lnurl? for what type of payment method did you see this behavior?
> 
> the updates you made seems great given the highlighted issue, but that also make a code a bit more complex to comprehend, so I'm wondering if we don't have an issue otherwise in our code. I don't think it should take long to validate an invoice for instance

Here's a video from main branch. Note the time from the clipboard button being opaque from being pressed. It's about 2.25 seconds until there's a response. And this wait would be longer with poor internet connection.
https://github.com/GaloyMoney/galoy-mobile/assets/51105802/2c9fa7e3-0dcc-4c8b-bb56-0e76682ab6e3

I know it's not the end of the world, but yea. I found it annoying and felt the need to improve it since I was already working on the same component. Here's the speed now:

https://github.com/GaloyMoney/galoy-mobile/assets/51105802/b2672a67-70d2-47bc-9e80-6a7c51f78f19

**Comment by MaxwellDG** [comment]
> > theres a multiple seconds lag on pressing the paste button for the first time which is very uncomfortable. Gonna work on making that more snappy. Likely it was only happening on Android
> 
> I have not seen this behavior AFAIR, maybe expect with lnurl? for what type of payment method did you see this behavior?
> 
> the updates you made seems great given the highlighted issue, but that also make a code a bit more complex to comprehend, so I'm wondering if we don't have an issue otherwise in our code. I don't think it should take long to validate an invoice for instance

From the home-screen I'm pressing 'send' and then inputting a btc address.

Here's a video from main branch. Note the time from the clipboard button being opaque from being pressed. It's about 2.25 seconds until there's a response. And this wait would be longer with poor internet connection.

https://github.com/GaloyMoney/galoy-mobile/assets/51105802/86bdaff3-3e8b-4632-8401-346d5e058425

I know it's not the end of the world, but yea. I found it annoying and felt the need to improve it since I was already working on the same component. Here's the speed now:

https://github.com/GaloyMoney/galoy-mobile/assets/51105802/501783ec-0da4-45b7-ab9d-4e33a51640e0

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx (hunk 3 above)
Comment (MaxwellDG) [reply app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:293]: yea was overkill. removed it.

File: app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx (hunk 1 above)
Comment (MaxwellDG) [reply app/screens/send-bitcoin-screen/send-bitcoin-destination-screen.tsx:162]: I've removed them in most instances, but I left them there with a comment why there's two checks. 1 before changing state 'validating', and 1 before actually doing validation.

**Comment by MaxwellDG** [comment]
> > lnbc10n1pjh09c7pp5cxgqx8ptspl28s43pxt20aae40pcqa3fx86uyyzng2vhup8j0uesdq5g9kxy7fqd9h8vmmfvdjscqzzsxqyz5vqsp5dj42yuzldlf0r4yee5gn3qlctq8tqkp8jqrt77yc38vdgnmeksks9qyyssqt558gadxwp95y0f75n6hrz5mgtf3wtg93xz04jmf4tuatyh8kpyn66vs20ug64xwcwxqsmxe4hjer2utlkjx0fnhpvndvlqlqemq94gq6ecx7l
> 
> Found the prop that fixes that. Ready for a real review now. Weird test not passing. Don't think it's related to this PR.

Hmmm this isn't gonna work actually. It messes up the normal typing process cause the 'selection' selects the whole string while typing. I'll have to think of something else tmrw.

**Comment by nicolasburtey** [comment]
> I know it's not the end of the world, but yea. I found it annoying and felt the need to improve it since I was already working on the same component. Here's the speed now:

ok that's better indeed. it didn't bother me really but if it's bothering you it's probably bothering other, so that's a great improvement thank you.

I'm not sure about the loading indicator on the paste button though, ie here:

image

- there are multiple ways to paste, and here it feels this may have this behavior only if this button is pushed
- there is already a loading indicator on the send button at the bottom
- we have many ways to load a string on this screen. paste / type / scan. I think it's better to be consistent in how we show the loading indicator the send button at the bottom feels like the most appropriate way to be consistent across all way to add the destination string

**Comment by MaxwellDG** [comment]
closing in order to open two separate PRs that are more targeted to their related task
